pvlib/iam.py

```diff
@@ -274,6 +274,119 @@ def martin_ruiz(aoi, a_r=0.16):
+        This is an empirical dimensionless parameter. Values of a_r are
+        generally on the order of 0.08 to 0.25 for flat-plate PV modules.
+        a_r must be greater than zero.
+
+    c1 : numeric scalar
+        First fitting parameter for the expressions that approximates the
+        integral of diffuse irradiance coming from different directions.
+        c1 is given as the constant 4 / 3 / pi (0.4244) in [1].
+
+    c2 : numeric scalar
+        Second fitting parameter for the expressions that approximates the
+        integral of diffuse irradiance coming from different directions.
+        If c2 is None, it will be calculated according to the linear
+        relationship given in [3].
+
+    Returns
+    -------
+    iam_sky : numeric
+        The incident angle modifier for sky diffuse
+
+    iam_ground : numeric
+        The incident angle modifier for ground-reflected diffuse
+
+    Notes
+    -----
+    Sky and ground modifiers are complementary: iam_sky for tilt = 30 is
+    equal to iam_ground for tilt = 180 - 30.  For vertical surfaces,
+    tilt = 90, the two factors are equal.
+
+    References
+    ----------
+    [1] N. Martin and J. M. Ruiz, "Calculation of the PV modules angular
+    losses under field conditions by means of an analytical model", Solar
+    Energy Materials & Solar Cells, vol. 70, pp. 25-38, 2001.
+
+    [2] N. Martin and J. M. Ruiz, "Corrigendum to 'Calculation of the PV
+    modules angular losses under field conditions by means of an
+    analytical model'", Solar Energy Materials & Solar Cells, vol. 110,
+    pp. 154, 2013.
+
+    [3] "IEC 61853-3 Photovoltaic (PV) module performance testing and energy
+    rating - Part 3: Energy rating of PV modules". IEC, Geneva, 2018.
+
+    See Also
+    --------
+    iam.martin_ruiz
+    iam.physical
+    iam.ashrae
+    iam.interp
+    iam.sapm
+    '''
+    # Contributed by Anton Driesse (@adriesse), PV Performance Labs. Oct. 2019
+
+    if isinstance(surface_tilt, pd.Series):
+        out_index = surface_tilt.index
+    else:
+        out_index = None
+
+    surface_tilt = np.asanyarray(surface_tilt)
+
+    # undo possible surface rotations
+    with np.errstate(invalid='ignore'):
+        surface_tilt = (surface_tilt + 180) % 360 - 180
+
+    # flip backward tilts forward
+    surface_tilt = np.abs(surface_tilt)
+
+    # avoid undefined results for horizontal or upside-down surfaces
+    small_angle = 1e-06
+    surface_tilt = np.clip(surface_tilt, small_angle, 180 - small_angle)
+
+    if np.any(np.less_equal(a_r, 0)):
+        raise ValueError("The parameter 'a_r' cannot be zero or negative.")
+
+    if c2 is None:
+        # This equation is from [3] Sect. 7.2
+        c2 = 0.5 * a_r - 0.154
+
+    from numpy import pi, radians, sin, cos, exp
+
+    beta = radians(surface_tilt)
+
+    with np.errstate(invalid='ignore'):
+        # because sin(pi) isn't exactly zero
+        sin_beta = np.where(surface_tilt < 90, sin(beta), sin(pi - beta))
```

Comment:
Since line 360 limits `surface_tilt` between 0 and 180, this line is unnecessary.

Reply:
This is a numeric precision issue.

Reply:
Since sin(theta) = sin( pi - theta) I don't see what this line accomplishes.

Reply:
It ensures that when surface_tilt is 180, sin_beta is exactly 0.

Reply:
```
In [37]: np.sin(np.pi)
Out[37]: 1.2246467991473532e-16
```

Aha.